File: SinglyCircular_LL/SinglyCL1.cpp
#include <iostream>
using namespace std;

struct node
{
    int data;
    struct node *next;
};

typedef struct node NODE;
typedef struct node *PNODE;

class SinglyCL
{
public:
    PNODE first;
    PNODE last;
    int iCount;

    SinglyCL()
    {
        cout << "Objects of singlyCL gets created :" << "\n";
        first = NULL;
        last = NULL;
        iCount = 0;
    }

    void InsertFirst(int no)
    {
        PNODE newn = NULL;

        newn = new NODE;

        newn->data = no;
        newn->next = NULL;

        if ((first == NULL) && (last == NULL))
        {
            first = newn;
            last = newn;
        }
        else
        {
            newn->next = first;
            first = newn;
        }


        last->next = first;
        iCount++;
    }
    void InsertLast(int no)
    {
        PNODE newn = NULL;
        PNODE temp = NULL;

        newn = new NODE;

        newn->data = no;
        newn->next = NULL;

        if ((first == NULL) && (last == NULL))
        {
            first = newn;
            last = newn;
        }
        else
        {
            last->next = newn;
            last = newn;

        }

        last->next = first;
        iCount++;
    }
    void DeleteFirst()
    {
    }
    void DeleteLast()
    {
    }
    void Display()
    {
        PNODE temp = NULL;

        temp = first;

        do
        {
            cout << "| " << temp->data << " |=>";
            temp = temp->next;

        }while(temp != last->next);

        cout << "\n";
    }

    int Count()
    {
        return iCount;
    }

    void InsertAtPos(int no, int pos)
    {
    }

    void DeleteAtPos(int pos)
    {
    }
};

int main()
{
    SinglyCL obj;

    int iRet = 0;

    obj.InsertFirst(51);
    obj.InsertFirst(41);
    obj.InsertFirst(31);
    obj.InsertFirst(21);
    obj.InsertFirst(11);

    obj.Display();

    iRet = obj.Count();

    cout << "Number of nodes are :" << iRet << "\n";

    obj.InsertLast(101);
    obj.InsertLast(111);

    obj.Display();

    iRet = obj.Count();

    cout << "Number of nodes are :" << iRet << "\n";

    return 0;
}

File: SinglyCircular_LL/SinglyCL1.py
class Node:
    def __init__(self,data):
        self.data = data
        self.next = None

class SinglyCL:
    def __init__(self):
        self.first = None
        self.last = None
        self.iCount = 0

    def InsertFirst(self,no):
       
        newn = Node(no)

        if((self.first == None) and (self.last == None)):

            self.first = newn
            self.last = newn

        else:
            newn.next = self.first
            self.first = newn


        self.last.next = self.first
        self.iCount = self.iCount + 1

    def InsertLast(self,no):

        newn = Node(no)

        if((self.first == None) and (self.last == None) ):

            self.first = newn
            self.last = newn

        else:
            
            self.last.next = newn
            self.last = newn


        self.last.next = self.first
        self.iCount = self.iCount + 1


    def DeleteFirst(self):
        pass

    def DeleteLast(self):
        pass

    def Display(self):

        if self.first is None:
            print("Linked List is empty")
            return
        
        temp = self.first

        while True:
            print("| ", temp.data, " |->", end=" ")
            temp = temp.next

            if temp == self.last.next:
                break

        print()

    def Count(self):
        return self.iCount

    def InsertAtPosFirst(self,no,pos):
        pass


    def DeleteAtPos(self,pos):
        pass


def main():

    sobj = SinglyCL()

    sobj.InsertFirst(101)
    sobj.InsertFirst(51)
    sobj.InsertFirst(21)
    sobj.InsertFirst(11)

    print("Elements of Linked List are :")

    sobj.Display()
    print("Number of Elements in Linekd List are :",sobj.iCount)

    print()

    sobj.InsertLast(101)
    sobj.InsertLast(111)

    print("Elements of Linked List are :")

    sobj.Display()
    print("Number of Elements in Linekd List are :",sobj.iCount)

    print()

if __name__ == "__main__":
    main()